src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs

```diff
@@ -385,22 +401,44 @@ private void UpdateReferences(TransactionOperationContext indexContext, string c
             }
         }
 
+        private static void RenewTransactionIfNeeded(QueryOperationContext queryContext, Index.CanContinueBatchResult batchContinuationResult, HashSet<Slice> indexed, ref IDisposable readTransaction)
+        {
+            if (batchContinuationResult != Index.CanContinueBatchResult.RenewTransaction)
+                return;
+
+            var toDispose = readTransaction;
+            readTransaction = null; // prevent double dispose in case of an error
+            toDispose.Dispose();
+
+            foreach (var slice in indexed)
+            {
+                slice.Release(queryContext.Documents.Allocator);
+            }
+
+            // the documents that have already been indexed become irrelevant once the new transaction is opened.
+            indexed.Clear();
+
+            readTransaction = queryContext.OpenReadTransaction();
+        }
+
         private IEnumerable<IndexItem> GetItemsFromCollectionThatReference(QueryOperationContext queryContext, TransactionOperationContext indexContext,
-            string collection, Reference referencedItem, long lastIndexedEtag, HashSet<string> indexed, ReferencesState.ReferenceState referenceState)
+            string collection, Reference referencedItem, long lastIndexedEtag, HashSet<Slice> indexed, ReferencesState.ReferenceState referenceState)
         {
             var lastProcessedItemId = referenceState?.GetLastProcessedItemId(referencedItem);
             foreach (var key in _referencesStorage.GetItemKeysFromCollectionThatReference(collection, referencedItem.Key, indexContext.Transaction, lastProcessedItemId))
             {
-                var item = GetItem(queryContext.Documents, key);
-                if (item == null)
-                    continue;
-
-                if (indexed.Add(item.Id) == false)
+                // we check if we already indexed the document BEFORE we fetch it from the disk.
+                var clonedKey = key.Clone(queryContext.Documents.Allocator);
```

Comment:
Instead of always doing `Clone()` you could do the check with the usage of `Contains()`. If it's not there then you'll do `Clone()` and `indexed.Add()`

Reply:
done